### PR TITLE
[all_tests] Default timeout of 3 min for all pytest

### PR DESCRIPTION
**Pull request by rickyyx**
## Why are these changes needed?
On some slow mac OS machine, we have seen a single pytest (a test function, not the file) runs for more than 10min and timeouts, resulting the entire job to timeout. 

A py unit test should not be running this long given it's run on every commit, this value is set higher than expectation to allow minimal interruptions of current CI, but catch abnormally. 

Mac test now timeouts much faster (instead of running 10min+) on an eventually-going-to-fail test with explicit stack trace: 
image

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: pytest.ini [hunk 1]
```diff
@@ -8,3 +8,6 @@ filterwarnings =
 
     # The format is `action:message_regex:category:module:line`.
     ignore:.*:
+
+# Default to 3 min timeout for all individual pytest.
+timeout = 180
```
Comment (aslonnie) [inline pytest.ini:13]: I do not think this affects bazel tests?

**Review by aslonnie — COMMENTED** [review]
do you have a link to run in concern (rather than the screenshot)?

I think timeout limiting should be done in bazel `--test_timeout` flag rather than in pytest, and one should adjust the timeout or size attribute of a bazel pytest rule to provide different timeout values for different tests.

**Comment by rickyyx** [comment]
> do you have a link to run in concern (rather than the screenshot)?

**Status Quo**
Oh yes, this is one of the current run that times out only at module level:  https://buildkite.com/ray-project/oss-ci-build-pr/builds/34044#018a3d71-843f-46d7-8a5a-f37ddb52e2dc

See the timestamp (15min = 21:45:52 -> 22:00:09) 
image


**With this PR** 
A timeout of a single pytest will be raised with traceback of threads so one could actually know where the test driver is stuck at, see https://buildkite.com/ray-project/oss-ci-build-pr/builds/33844#018a2efb-e4f6-4fd6-b102-92f13dcbc8b8

image




> I think timeout limiting should be done in bazel `--test_timeout` flag rather than in pytest, and one should adjust the timeout or size attribute of a bazel pytest rule to provide different timeout values for different tests.

I think the reason to enforce a per pytest timeout is that a signle python test module might contain multiple pytest test functions, while the bazel flag set the timeout for the entire file. I feel setting timeout on pytest is also needed since this provides isolation of individual tests within the same module. 

There's no point waiting for a single pytest to eat up the entire module timeout time since a module could contain quite a number of pytest tests. 


> I do not think this affects bazel tests?

This has an impact: see this: https://buildkite.com/ray-project/oss-ci-build-pr/builds/33844#018a2eea-3261-47ac-9ba3-e32d7f3d9ba2

**Comment by rickyyx** [comment]
Also, TIL, we apparently change timeout based on different bazel flags than the default: https://github.com/ray-project/ray/blob/6deefe6c22026d31a73c77e79fc5522f7f5c3f7d/ci/ray_ci/bazel_sharding.py#L48-L66

**Comment by can-anyscale** [comment]
This looks fine to me, but there are some timedout tests on CI https://buildkite.com/ray-project/oss-ci-build-pr/builds/34036#018a3d3e-9037-4edf-94b9-6cc3e03969b9 and I cannot tell if they are relevant or not

**Comment by rickyyx** [comment]
> This looks fine to me, but there are some timedout tests on CI https://buildkite.com/ray-project/oss-ci-build-pr/builds/34036#018a3d3e-9037-4edf-94b9-6cc3e03969b9 and I cannot tell if they are relevant or not

It's probably transient. Retry passes.

**Review by can-anyscale — APPROVED** [review]

**Pull request merged**